data/pkmn_sets.py

```diff
@@ -38,7 +43,80 @@
     from fp.battle import Pokemon
 
 logger = logging.getLogger(__name__)
-PWD = os.path.dirname(os.path.abspath(__file__))
+
+
+def get_sets_file(cache_path: str, remote_url: str) -> dict:
+    if os.path.exists(cache_path):
+        with open(cache_path, "r") as f:
+            sets = json.load(f)
+        logger.info(f"Loaded from cache: {cache_path}")
+        return sets
+
+    r = requests.get(remote_url)
+    if r.status_code == 200:
+        sets = r.json()
+    else:
+        logger.warning(
+            f"Could not retrieve from remote: {remote_url} "
+            f"(status code {r.status_code})"
+        )
+        sets = {}
+
+    os.makedirs(os.path.dirname(cache_path), exist_ok=True)
+    with open(cache_path, "w") as f:
+        json.dump(sets, f)
+    logger.info(f"Downloaded and cached from remote: {remote_url}")
+    return sets
+
+
+def get_ps_sets_file(pkmn_mode: str) -> dict:
+    def add_new_set(d: dict, p: str, s: str):
+        if p not in d:
+            d[p] = {}
+        if s not in d[p]:
+            d[p][s] = 0
+        d[p][s] += 1
+
+    def get_pkmn_data(n: dict, s: dict):
+        for pkmn_name, pkmn_sets in s.items():
+            pkmn_name = normalize_name(pkmn_name)
+            for pkmn_set in pkmn_sets.values():
+                moves = "|".join(sorted([normalize_name(m) for m in pkmn_set["moves"]]))
+                ability = normalize_name(pkmn_set["ability"])
+                item = normalize_name(pkmn_set["item"])
+                nature = normalize_name(pkmn_set["nature"])
+                tera_type = normalize_name(pkmn_set.get("teraType", ""))
+                evs = (
+                    pkmn_set["evs"].get("hp", 0),
+                    pkmn_set["evs"].get("atk", 0),
+                    pkmn_set["evs"].get("def", 0),
+                    pkmn_set["evs"].get("spa", 0),
+                    pkmn_set["evs"].get("spd", 0),
+                    pkmn_set["evs"].get("spe", 0),
+                )
+                evs = ",".join(str(v) for v in evs)
```

Comment:
todo: this won't work for all past generations, e.g. [gen1ou.json](https://play.pokemonshowdown.com/data/sets/gen1ou.json) will cause a KeyError here for ability/item/nature/evs.

Pls fix